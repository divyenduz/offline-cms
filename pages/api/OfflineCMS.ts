import cheerio from 'cheerio'
import prettier from 'prettier'

export class OfflineCMS {
  private readonly $: CheerioStatic

  constructor(content: string) {
    this.$ = cheerio.load(content)
  }

  prettify() {
    return prettier.format(this.getContent(), {
      parser: 'html',
    })
  }

  editBody(newBody) {
    const $ = this._getClone() // Clone it
    console.log(newBody)
    $('body').html(newBody)
    return new OfflineCMS($.html())
  }

  getShell() {
    const $ = this._getClone() // Clone it
    $('body').html('')
    return $.html()
  }

  getBody() {
    return this.$('body').html()
  }

  getContent() {
    return this.$.html()
  }

  private _getClone() {
    return cheerio.load(this.$.html())
  }
}
