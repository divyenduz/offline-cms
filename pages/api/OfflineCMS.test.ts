import { identityTag as html } from 'identity-tag'
import { OfflineCMS } from './OfflineCMS'

const content = html`
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="UTF-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <title>Future</title>
      <style>
        .one {
          display: block;
        }
      </style>
      <style>
        .two {
          display: block;
        }
      </style>
    </head>
    <body>
      <ul>
        <li>
          Publish offline-cms to
          <a href="https://github.com/divyenduz/offline-cms">Github</a> âœ…
        </li>
        <li>Publish offline-cms tweet</li>
      </ul>
    </body>
  </html>
`

describe('offline cms', () => {
  it('should return the shell', () => {
    const offlineCMS = new OfflineCMS(content)
    expect(offlineCMS.getShell()).toMatchSnapshot()
  })

  it('should return the body', () => {
    const offlineCMS = new OfflineCMS(content)
    expect(offlineCMS.getBody()).toMatchSnapshot()
  })

  it('should return the content', () => {
    const offlineCMS = new OfflineCMS(content)
    expect(offlineCMS.getContent()).toMatchSnapshot()
  })

  it('should edit the body', () => {
    const offlineCMS = new OfflineCMS(content)
    expect(offlineCMS.editBody('<p>Hello</p>').getContent()).toMatchSnapshot()
  })

  it('should prettify', () => {
    const offlineCMS = new OfflineCMS(content)
    expect(offlineCMS.prettify()).toMatchSnapshot()
  })

  it('should prettify after edit', () => {
    const offlineCMS = new OfflineCMS(content)
    expect(offlineCMS.editBody('<p>Hello</p>').prettify()).toMatchSnapshot()
  })

  it('should get styles', () => {
    const offlineCMS = new OfflineCMS(content)
    expect(offlineCMS.getStyles()).toMatchSnapshot()
  })
})
