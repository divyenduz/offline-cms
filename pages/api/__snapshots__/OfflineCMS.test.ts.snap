// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offline cms should edit the body 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
      <meta charset=\\"UTF-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
      <title>Future</title>
    </head>
    <body><p>Hello</p></body></html>"
`;

exports[`offline cms should prettify 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Future</title>
  </head>
  <body>
    <ul>
      <li>
        Publish offline-cms to
        <a href=\\"https://github.com/divyenduz/offline-cms\\">Github</a> &#x2705;
      </li>
      <li>Publish offline-cms tweet</li>
    </ul>
  </body>
</html>
"
`;

exports[`offline cms should prettify after edit 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Future</title>
  </head>
  <body>
    <p>Hello</p>
  </body>
</html>
"
`;

exports[`offline cms should return the body 1`] = `
"
      <ul>
        <li>
          Publish offline-cms to
          <a href=\\"https://github.com/divyenduz/offline-cms\\">Github</a> &#x2705;
        </li>
        <li>Publish offline-cms tweet</li>
      </ul>
    
  
"
`;

exports[`offline cms should return the content 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
      <meta charset=\\"UTF-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
      <title>Future</title>
    </head>
    <body>
      <ul>
        <li>
          Publish offline-cms to
          <a href=\\"https://github.com/divyenduz/offline-cms\\">Github</a> &#x2705;
        </li>
        <li>Publish offline-cms tweet</li>
      </ul>
    
  
</body></html>"
`;

exports[`offline cms should return the shell 1`] = `
"<!DOCTYPE html><html lang=\\"en\\"><head>
      <meta charset=\\"UTF-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
      <title>Future</title>
    </head>
    <body></body></html>"
`;
